### PR TITLE
[aggregator] Avoid large copies in entry rollup comparisons by making them more inline-friendly

### PR DESCRIPTION
**Pull request by vdarulis**
**What this PR does / why we need it**:

Turns out Go was unable to inline most of the comparison logic, adding function call and copy overhead for metadata:
```
name                  old time/op    new time/op    delta
AggregationValues-12     105ns ± 0%      43ns ± 1%  -58.70%  (p=0.000 n=8+8)
```

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note

```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note

```

### CODE REVIEW

File: src/aggregator/aggregator/entry.go [hunk 1]
```diff
@@ -1076,8 +1076,14 @@ type aggregationValue struct {
 type aggregationValues []aggregationValue
 
 func (vals aggregationValues) index(k aggregationKey) int {
-	for i, val := range vals {
-		if val.key.Equal(k) {
+	// keep in sync with aggregationKey.Equal()
+	// this is >2x slower if not inlined manually.
+	for i := range vals {
```
Comment (andrewmains12) [inline src/aggregator/aggregator/entry.go:1081]: Hmm dunno if this is possible, but any way to add a fuzz test comparing behavior using equals vs using the manual inline here? Less to test the current version of the code than to ensure against future regressions haha.

File: src/aggregator/aggregator/forwarded_writer.go [hunk 2]
```diff
@@ -474,8 +474,8 @@ func (agg *forwardedAggregation) onDone(key aggregationKey) error {
 }
 
 func (agg *forwardedAggregation) index(key aggregationKey) int {
-	for i, k := range agg.byKey {
-		if k.key.Equal(key) {
+	for i := range agg.byKey {
+		if agg.byKey[i].key.Equal(key) {
```
Comment (andrewmains12) [inline src/aggregator/aggregator/forwarded_writer.go:478]: Dumb question: this is faster? I'm surprised haha--more inlining issues?

File: src/aggregator/aggregator/entry.go (hunk 1 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/entry.go:1081]: Also, maybe worth adding a pointer from `Equal` to here to make sure anyone changing that method also changes this one?

Makes one wish for macros...

File: src/aggregator/aggregator/forwarded_writer.go (hunk 2 above)
Comment (mway) [reply src/aggregator/aggregator/forwarded_writer.go:478]: stack copies due to how range works

File: src/metrics/pipeline/applied/type.go [hunk 3]
```diff
@@ -91,25 +91,25 @@ type OpUnion struct {
 
 // Equal determines whether two operation unions are equal.
 func (u OpUnion) Equal(other OpUnion) bool {
+	// keep in sync with Pipeline.Equal as go is terrible at inlining anything with a loop
 	if u.Type != other.Type {
 		return false
 	}
-	switch u.Type {
-	case pipeline.TransformationOpType:
```
Comment (mway) [inline src/metrics/pipeline/applied/type.go:98]: what about this case?

File: src/metrics/pipeline/applied/type.go [hunk 4]
```diff
@@ -91,25 +91,25 @@ type OpUnion struct {
...
 // Clone clones an operation union.
 func (u OpUnion) Clone() OpUnion {
-	clone := OpUnion{Type: u.Type}
-	switch u.Type {
-	case pipeline.TransformationOpType:
-		clone.Transformation = u.Transformation.Clone()
-	case pipeline.RollupOpType:
+	clone := OpUnion{
+		Type:           u.Type,
+		Transformation: u.Transformation,
```
Comment (mway) [inline src/metrics/pipeline/applied/type.go:110]: does this need a clone as well? if so, two ifs is probably going to put you back in switch land branch wise

File: src/metrics/pipeline/applied/type.go [hunk 5]
```diff
@@ -187,11 +187,18 @@ func (p Pipeline) At(i int) OpUnion { return p.operations[i] }
...
 	}
+
 	for i := 0; i < len(p.operations); i++ {
-		if !p.operations[i].Equal(other.operations[i]) {
+		if p.operations[i].Type != other.operations[i].Type {
+			return false
+		}
+
+		if p.operations[i].Type == pipeline.RollupOpType &&
+			!p.operations[i].Rollup.Equal(other.operations[i].Rollup) {
```
Comment (mway) [inline src/metrics/pipeline/applied/type.go:201]: do we need to worry about the transform case?

File: src/metrics/pipeline/applied/type.go (hunk 3 above)
Comment (vdarulis) [reply src/metrics/pipeline/applied/type.go:98]: The definition is `type TransformationOp struct{Type transformation.Type}`, where `transformation.Type` is plain `int`

File: src/metrics/pipeline/applied/type.go (hunk 4 above)
Comment (vdarulis) [reply src/metrics/pipeline/applied/type.go:110]: nope, copying an int > copying the same single int conditionally

File: src/aggregator/aggregator/entry.go (hunk 1 above)
Comment (vdarulis) [reply src/aggregator/aggregator/entry.go:1081]: good point as I was just lazy, but then again I missed something in tests :D

File: src/aggregator/aggregator/forwarded_writer.go (hunk 2 above)
Comment (andrewmains12) [reply src/aggregator/aggregator/forwarded_writer.go:478]: Ah nice, TIL.

**Review by mway — APPROVED** [review]

File: src/metrics/pipeline/applied/type.go (hunk 5 above)
Comment (vdarulis) [reply src/metrics/pipeline/applied/type.go:201]: yes, my bad, missed it

**Pull request merged**